enterprise/server/iprules/iprules.go

```diff
@@ -0,0 +1,261 @@
+		return nil, false
+	}
+	return entry.allowed, true
+}
+
+func (c *memIpRuleCache) Add(groupID string, allowed []*net.IPNet) {
+	c.mu.Lock()
+	c.lru.Add(groupID, &ipRuleCacheEntry{allowed: allowed, expiresAfter: time.Now().Add(*cacheTTL)})
+	c.mu.Unlock()
+}
+
+type noopIpRuleCache struct {
+}
+
+func (c *noopIpRuleCache) Add(groupID string, allowed []*net.IPNet) {
+}
+
+func (c *noopIpRuleCache) Get(groupID string) ([]*net.IPNet, bool) {
+	return nil, false
+}
+
+func newIpRuleCache() (ipRuleCache, error) {
+	if *cacheTTL == 0 {
+		return &noopIpRuleCache{}, nil
+	}
+	config := &lru.Config[*ipRuleCacheEntry]{
+		MaxSize: cacheSize,
+		SizeFn:  func(v *ipRuleCacheEntry) int64 { return int64(len(v.allowed)) },
+	}
+	l, err := lru.NewLRU[*ipRuleCacheEntry](config)
+	if err != nil {
+		return nil, err
+	}
+	return &memIpRuleCache{
+		lru: l,
+	}, nil
+}
+
+type Service struct {
+	env environment.Env
+
+	cache ipRuleCache
+}
+
+func New(env environment.Env) (*Service, error) {
+	cache, err := newIpRuleCache()
+	if err != nil {
+		return nil, err
+	}
+	return &Service{
+		env:   env,
+		cache: cache,
+	}, nil
+}
+
+func Register(env environment.Env) error {
+	if !*enableIPRules {
+		return nil
+	}
+
+	enforcer, err := New(env)
+	if err != nil {
+		return err
+	}
+	env.SetIPRulesService(enforcer)
+	return nil
+}
+
+func (s *Service) loadRulesFromDB(ctx context.Context, groupID string) ([]*net.IPNet, error) {
+	rows, err := s.env.GetDBHandle().DB(ctx).Raw(
+		`SELECT * FROM "IPRules" WHERE group_id = ?`, groupID).Rows()
+	if err != nil {
+		return nil, err
+	}
+
+	var allowed []*net.IPNet
+	for rows.Next() {
+		r := &tables.IPRule{}
+		if err := s.env.GetDBHandle().DB(ctx).ScanRows(rows, r); err != nil {
+			return nil, err
+		}
+		_, ipNet, err := net.ParseCIDR(r.CIDR)
+		if err != nil {
+			alert.UnexpectedEvent("unparsable CIDR rule", "rule %q", r.CIDR)
+			continue
+		}
+		allowed = append(allowed, ipNet)
+	}
+	return allowed, nil
+}
+
+func (s *Service) AuthorizeGroup(ctx context.Context, groupID string) error {
+	g, err := s.env.GetUserDB().GetGroupByID(ctx, groupID)
+	if err != nil {
+		return err
+	}
+	if !g.EnforceIPRules {
+		return nil
+	}
+
+	rawClientIP := clientip.Get(ctx)
+	clientIP := net.ParseIP(rawClientIP)
+	// Client IP is not parsable.
+	if clientIP == nil {
+		return status.FailedPreconditionErrorf("client IP %q is not valid", rawClientIP)
+	}
+
+	allowed, ok := s.cache.Get(groupID)
+	if !ok {
+		rs, err := s.loadRulesFromDB(ctx, groupID)
+		if err != nil {
+			return err
+		}
+		s.cache.Add(groupID, rs)
+		allowed = rs
+	}
+
+	for _, a := range allowed {
+		if a.Contains(clientIP) {
+			return nil
+		}
+	}
+
+	return status.PermissionDeniedErrorf("Client %q is not allowed by Organization IP rules", rawClientIP)
+}
+
+func (s *Service) Authorize(ctx context.Context) error {
+	u, err := perms.AuthenticatedUser(ctx, s.env)
+	if authutil.IsAnonymousUserError(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
```

Comment:
Idk if this is much better but you could.
```
	if err != nil {
    	if authutil.IsAnonymousUserError(err) {
	    	return nil
    	}
		return err
	}
```

Do we want anon users returning no error here?

Reply:
For anon users, we don't have a group to check against. Anon users will be handled by regular auth checks.